porcupine/actions.py

```diff
@@ -0,0 +1,140 @@
+from dataclasses import dataclass
+from functools import partial
+from pathlib import Path
+from typing import Callable
+
+from porcupine.tabs import FileTab
+
+action_availability_callback = Callable[[], bool]
+
+
+@dataclass(kw_only=True, slots=True, frozen=True)
+class Action:
+    """Action that requires no context in the callback"""
+
+    name: str
+    description: str
+    callback: Callable[..., None]
+    availability_callbacks: list[action_availability_callback] | None = None
+
+
+filetab_action_availability_callback = Callable[[FileTab], bool]
+
+
+@dataclass(kw_only=True, slots=True, frozen=True)
+class FileTabAction:
+    """Action that requires a FileTab to be provided to the callback"""
+
+    name: str
+    description: str
+    callback: Callable[[FileTab], None]
+    availability_callbacks: list[filetab_action_availability_callback] | None = None
+
+
+path_action_availability_callback = Callable[[Path], bool]
+
+
+@dataclass(kw_only=True, slots=True, frozen=True)
+class PathAction:
+    """Action that requires a Path to be provided to the callback"""
+
+    name: str
+    description: str
+    callback: Callable[[Path], None]
+    file_compatible: bool = True
+    directory_compatible: bool = False
+    availability_callbacks: list[path_action_availability_callback] | None = None
+
+
+ActionTypes = Action | FileTabAction | PathAction
```

Comment:
IMO this union should be named `Action`. You may need to use the `Union[Foo, Bar, Baz]` syntax here to support some of the older Python versions we test in CI.

Reply:
does `from __future__ import Anotations` not solve that?

Reply:
As far as the runtime is concerned, this is an assignment, `foo = bar`. The right side must be evaluated. The `__future__` thing helps only with `:` syntax, as in `foo: bar`.

Reply:
Yeah I see that that's correct. What versions of Python does Porcupine support?

Reply:
3.8 and newer, though pyls is terrible on 3.11.

Reply:
this is fixed